File: spring-data/example/src/main/java/dev/cerbos/example/demo/DemoApplication.java
/*
 * Copyright 2021-2026 Zenauth Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

package dev.cerbos.example.demo;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

import dev.cerbos.example.CerbosClientConfig;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.Banner;
import org.springframework.boot.WebApplicationType;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.builder.SpringApplicationBuilder;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Import;

import java.io.PrintStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;

/**
 * The demo-domain program run by {@code demo/scripts/run-example.sh spring-data}.
 *
 * <p>It runs the five shared usage shapes against {@code demo/}'s policies and seed rows and
 * prints one JSON document to stdout, which the runner diffs against {@code demo/expected.json}.
 * Component scanning covers only this package, so the photo-sharing entities are not registered.
 */
@SpringBootApplication
@Import(CerbosClientConfig.class)
public class DemoApplication {

    // The real stdout. main() sends System.out to stderr so logs and stray prints cannot mix
    // with the JSON document, which is written through this handle.
    private static final PrintStream STDOUT = System.out;

    public static void main(String[] args) throws Exception {
        // Redirect and validate before Spring starts, so a misinvocation fails with one clear
        // message.
        System.setOut(System.err);
        Path demoDir = demoDir();
        requireCerbosHost();

        try (ConfigurableApplicationContext context = new SpringApplicationBuilder(DemoApplication.class)
                .web(WebApplicationType.NONE)
                .bannerMode(Banner.Mode.OFF)
                .profiles("demo")
                .properties("demo.seeds=" + demoDir.resolve("seeds.json"))
                .run(args)) {

            Map<String, Object> document = Map.of(
                    // The runner checks this name, so output from the wrong example fails.
                    "adapter", "spring-data",
                    "shapes", context.getBean(DemoShapes.class).run());

            // Map.of has no stable iteration order, so sort keys to keep the output readable
            // and stable between runs. The runner compares with `jq -S` either way.
            STDOUT.println(new ObjectMapper()
                    .enable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS)
                    .writerWithDefaultPrettyPrinter()
                    .writeValueAsString(document));
            STDOUT.flush();
        }
    }

    // The shared demo/ directory, passed as -Ddemo.dir by run.sh.
    private static Path demoDir() {
        String configured = System.getProperty("demo.dir");
        if (configured == null || configured.isBlank()) {
            throw new IllegalStateException(
                    "-Ddemo.dir is not set — run this program through"
                            + " demo/scripts/run-example.sh spring-data");
        }
        Path demoDir = Path.of(configured);
        if (!Files.isRegularFile(demoDir.resolve("seeds.json"))) {
            throw new IllegalStateException(
                    "-Ddemo.dir=" + configured + " does not contain seeds.json");
        }
        return demoDir;
    }

    // No default address: Cerbos's default ports may belong to another local PDP with other
    // policies, and planning against it would look like an adapter bug.
    private static void requireCerbosHost() {
        String host = System.getenv("CERBOS_HOST");
        if (host == null || host.isBlank()) {
            throw new IllegalStateException(
                    "CERBOS_HOST is not set — run this program through"
                            + " demo/scripts/run-example.sh spring-data");
        }
    }

    @Bean
    DemoSeeds demoSeeds(@Value("${demo.seeds}") String seedsFile) {
        return DemoSeeds.read(Path.of(seedsFile));
    }
}

File: exposed/example/src/main/kotlin/dev/cerbos/example/demo/DemoApplication.kt
@file:JvmName("DemoApplication")

package dev.cerbos.example.demo

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature
import dev.cerbos.sdk.CerbosBlockingClient
import dev.cerbos.sdk.CerbosClientBuilder
import java.io.PrintStream
import java.nio.file.Files
import java.nio.file.Path
import java.time.Duration
import org.jetbrains.exposed.v1.jdbc.Database
import org.jetbrains.exposed.v1.jdbc.transactions.transaction

/**
 * The real stdout, captured before [main] redirects `System.out`.
 *
 * The contract is one JSON document on stdout and everything else on stderr. Exposed, gRPC, SLF4J
 * and Jackson all sit on this classpath and any of them is free to print without a logging
 * configuration. Rather than configuring each of them, stdout is redirected to stderr for the
 * whole JVM and the one line that must reach the runner is written through this handle. It holds
 * under any logging configuration, and under a dependency that prints without one.
 *
 * A file-level property rather than a local: it is initialised when this file's class loads, which
 * is before the first statement of [main].
 */
private val REAL_STDOUT: PrintStream = System.out

/**
 * The exposed example: the program `demo/scripts/run-example.sh exposed` runs.
 *
 * It takes no arguments, exercises the
 * [five shared usage shapes](https://github.com/cerbos/query-plan-adapters/issues/349) against
 * `demo/`'s policies and seed rows, and prints exactly one JSON document to stdout for the shared
 * runner to diff against `demo/expected.json`.
 *
 * Two things it is handed, and where each comes from:
 *
 * - `CERBOS_HOST` — the environment variable the shared runner sets. There is deliberately no
 *   fallback; see [cerbosHost].
 * - `-Ddemo.dir` — the shared corpus directory. The path arithmetic lives in `run.sh`: how many
 *   directories up the repository root sits is the launcher's business, not the application's.
 *
 * The store is H2 in memory, inside this JVM, so there is nothing for `run.sh` to start.
 */
fun main() {
    // Before anything can print, and before anything can connect: a misinvocation costs one clear
    // message rather than a connection failure part-way through seeding.
    System.setOut(System.err)

    val demoDir = demoDir()
    val cerbos = cerbosClient(cerbosHost())
    val seeds = DemoSeeds.read(demoDir.resolve("seeds.json"))

    // `DB_CLOSE_DELAY=-1` keeps the in-memory database alive between connections, so the schema
    // this program creates survives for every query it then runs.
    val database = Database.connect(
        url = "jdbc:h2:mem:cerbos_exposed_demo;DB_CLOSE_DELAY=-1",
        driver = "org.h2.Driver",
    )

    // One transaction around every shape. The adapter returns an `Op<Boolean>`, a value that
    // renders inside the CALLER's transaction for that transaction's dialect, so the transaction
    // is the application's to open — which is the half of the contract an example should show.
    val shapes = transaction(database) { DemoShapes(cerbos, seeds).run() }

    val document = linkedMapOf(
        // The runner checks this against the adapter it was asked for, so a stale build directory
        // or a copied run.sh fails there instead of quietly passing on the shared expectations.
        "adapter" to "exposed",
        "shapes" to shapes,
    )

    // Map keys sorted, which is the same canonical form demo/scripts/run-example.sh puts both
    // sides of its diff into (`jq -S`), so the document a human reads on a failure is in the order
    // the diff is. Data class properties are unaffected and stay in declaration order.
    REAL_STDOUT.println(
        ObjectMapper()
            .enable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS)
            .writerWithDefaultPrettyPrinter()
            .writeValueAsString(document),
    )
    REAL_STDOUT.flush()
}

/**
 * Every call is bounded. A blocking gRPC stub with no deadline waits for ever, so a stalled stream
 * would hold this program, and the CI job running it, until something outside killed it. A real
 * application wants a deadline on its authorization calls for the same reason.
 */
private fun cerbosClient(host: String): CerbosBlockingClient =
    CerbosClientBuilder(host)
        .withPlaintext()
        .withTimeout(Duration.ofSeconds(30))
        .buildBlockingClient()

/** The shared corpus directory, passed as `-Ddemo.dir` by `run.sh`. */
private fun demoDir(): Path {
    val configured = System.getProperty("demo.dir")
    check(!configured.isNullOrBlank()) {
        "-Ddemo.dir is not set — run this program through demo/scripts/run-example.sh exposed"
    }
    val demoDir = Path.of(configured)
    check(Files.isRegularFile(demoDir.resolve("seeds.json"))) {
        "-Ddemo.dir=$configured does not contain seeds.json"
    }
    return demoDir
}

/**
 * The runner sets `CERBOS_HOST`, and there is deliberately no fallback anywhere in this example.
 * The obvious default — Cerbos's own 3592/3593 — is what every adapter's `cerbos run` test sidecar
 * binds, so an unset `CERBOS_HOST` would not fail: it would quietly plan against whichever policy
 * suite that sidecar serves, and produce a diff against `demo/expected.json` that reads as an
 * adapter bug. Two examples shipped that exact default, which is why
 * `demo/scripts/validate-demo.sh` now fails the build on a hardcoded PDP address rather than
 * trusting prose.
 */
private fun cerbosHost(): String {
    val host = System.getenv("CERBOS_HOST")
    check(!host.isNullOrBlank()) {
        "CERBOS_HOST is not set — run this program through demo/scripts/run-example.sh exposed"
    }
    return host
}
